src/text_shaper/directwrite_shaper.cpp

```diff
@@ -279,58 +284,137 @@ void directwrite_shaper::shape(font_key _font,
-    // auto actualGlyphCount = UINT32{0};
-
-    // analyzer->GetGlyphs(textString,
-    //                     textLength,
-    //                     fontFace,
-    //                     isSideways,
-    //                     isRightToLeft,
-    //                     scriptAnalysis,
-    //                     localeName, numberSubstitution,
-    //                     &features/*FIXME: is this right?*/,
-    //                     featureRangeLengths,
-    //                     featureRanges,
-    //                     maxGlyphCount,
-    //                     clusterMap,
-    //                     textProps,
-    //                     glyphIndices,
-    //                     glyphProps,
-    //                     &actualGlyphCount);
-
+    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> conv1;
+    std::string bytes = conv1.to_bytes(std::u32string{ _text });
+    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> conv2;
+    std::wstring wText = conv2.from_bytes(bytes);
+
+    WCHAR const* textString = wText.c_str();
+    UINT32 textLength = wText.size();
+    FontInfo const& fontInfo = d->fonts.at(_font);
+    IDWriteFontFace5* fontFace = fontInfo.fontFace.Get();
+
+    std::vector<UINT16> _glyphIndices;
+    std::vector<INT32> _glyphDesignUnitAdvances;
+
+    BOOL isTextSimple = FALSE;
+    UINT32 uiLengthRead = 0;
+    const UINT32 glyphStart = 0;
+    _glyphIndices.resize(textLength);
+
+    d->textAnalyzer->GetTextComplexity(
+        textString,
+        textLength,
+        fontFace,
+        &isTextSimple,
+        &uiLengthRead,
+        &_glyphIndices.at(glyphStart));
+
+    BOOL _isEntireTextSimple = isTextSimple && uiLengthRead == textLength;
+    if (_isEntireTextSimple)
+    {
+        DWRITE_FONT_METRICS1 metrics;
+        fontFace->GetMetrics(&metrics);
+
+        _glyphDesignUnitAdvances.resize(textLength);
+        USHORT designUnitsPerEm = metrics.designUnitsPerEm;
+        fontFace->GetDesignGlyphAdvances(
+            textLength,
+            & _glyphIndices.at(glyphStart),
+            & _glyphDesignUnitAdvances.at(glyphStart),
+            false);
+
+        for (size_t i = glyphStart; i < textLength; i++)
+        {
+            const auto cellWidth = static_cast<double>((float)_glyphDesignUnitAdvances.at(i)) / designUnitsPerEm * fontInfo.size.pt;
+            glyph_position gpos{};
+            gpos.glyph = glyph_key{ _font, fontInfo.size, glyph_index{_glyphIndices.at(i)} };
+            gpos.offset.x = static_cast<int>(cellWidth);
+            //gpos.offset.y = static_cast<int>(static_cast<double>(pos[i].y_offset) / 64.0f);
+
+            gpos.advance.x = static_cast<int>(cellWidth);
+            //gpos.advance.y = static_cast<int>(static_cast<double>(pos[i].y_advance) / 64.0f);
+            _result.emplace_back(gpos);
+        }
+    }
+    else {
+        // TODO
+    }
 }
 
 std::optional<rasterized_glyph> directwrite_shaper::rasterize(glyph_key _glyph, render_mode _mode)
```

Comment:
This is largely based on my previous experience and [crossfont](https://github.com/alacritty/crossfont).

One thing I don't understand is that, Alacritty doesn't do shaping at all. All it does in crossfont is `load_font` and then `rasterize`, which is why it doesn't support complex Emoji & ligatures. Here the `rasterize` in contour uses similar interface as in crossfont`(`glyph_key` -> `bitmap`). How does contour handles shaping? Sorry if this is a dumb question. I'm quite an amateur in this area.

Reply:
Yes. Alacritty does not text shaping at all because it is a very expensive task and does cost significant CPU (if done naive). I think one can at least save a bit via caching, which we do. :)

This rasterize function already takes a glyph ID regardless of whether text shaping was involved or a simple 1:1 translation has happened.

### How does contour handle shaping:

It does support two modes, basic shaping (just codepoint to glyph index mapping) and complex text shaping.

primitive codepoint-to-glyph-index translation is done via https://github.com/contour-terminal/contour/blob/master/src/text_shaper/shaper.h#L128-L129

whereas complex textshaping can be implemented via:https://github.com/contour-terminal/contour/blob/master/src/text_shaper/shaper.h#L109-L126

the open_shaper implementation does exactly that. the smaller of those two function where meant to bypass (complex) text shaping in order to gain performance (for the sake of losing features), but turns out in my naive tests it's actually not that much gained (due to caching) :-)

Reply:
WRT text shaping in terminals I can highly recommend reading https://github.com/contour-terminal/contour/blob/master/docs/text-stack.md - I've put a lot of afford into this document so I won't forget (or if I do, i have a place to quickly recap)

Reply:
Wonderful writeup about text shaping & rendering!